core/src/main/java/org/apache/brooklyn/core/OsgiLauncher.java

```diff
@@ -0,0 +1,557 @@
+import org.apache.brooklyn.core.server.BrooklynServerPaths;
+import org.apache.brooklyn.util.core.ResourceUtils;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.exceptions.FatalConfigurationRuntimeException;
+import org.apache.brooklyn.util.exceptions.FatalRuntimeException;
+import org.apache.brooklyn.util.exceptions.RuntimeInterruptedException;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.io.FileUtil;
+import org.apache.brooklyn.util.os.Os;
+import org.apache.brooklyn.util.stream.Streams;
+import org.apache.brooklyn.util.text.Strings;
+import org.apache.brooklyn.util.time.Duration;
+import org.apache.brooklyn.util.time.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Initializer for brooklyn-core when running in an OSGi environment.
+ *
+ * Temporarily here; should be totally contained in blueprint beans' init-methods.
+ */
+public class OsgiLauncher {
+
+    private static final Logger log = LoggerFactory.getLogger(OsgiLauncher.class);
+
+    private ManagementContext managementContext;
+
+    private boolean ignoreCatalogErrors = true;
+    private boolean ignorePersistenceErrors = true;
+
+    private HighAvailabilityMode highAvailabilityMode = HighAvailabilityMode.DISABLED;
+
+    private PersistMode persistMode = PersistMode.DISABLED;
+    private String persistenceDir = null;
+    private String persistenceLocation = null;
+    private Duration persistPeriod = Duration.ONE_SECOND;
+
+    public void start() {
+        // Configure launcher
+//        ShutdownHandler shutdownHandler = new AppShutdownHandler();
+        try {
+//            initBrooklynProperties();
+            initCatalog();
+
+            // TODO
+//            launcher.stopWhichAppsOnShutdown(stopWhichAppsOnShutdownMode);
+//            launcher.shutdownHandler(shutdownHandler);
+//
+//            computeAndSetApp(launcher, utils, loader);
+//            customize(launcher);
+//        } catch (FatalConfigurationRuntimeException e) {
+//            throw e;
+//        } catch (Exception e) {
+//            throw new FatalConfigurationRuntimeException("Fatal error configuring Brooklyn launch: " + e.getMessage(), e);
+//        }
+//
+//        try {
+            launcherStart();
+        } catch (FatalRuntimeException e) {
+            // rely on caller logging this propagated exception
+            throw e;
+        } catch (Exception e) {
+            // for other exceptions we log it, possibly redundantly but better too much than too little
+            Exceptions.propagateIfFatal(e);
+            log.error("Error launching brooklyn: " + Exceptions.collapseText(e), e);
+            try {
+                shutdown();
+            } catch (Exception e2) {
+                log.warn("Subsequent error during termination: " + e2);
+                log.debug("Details of subsequent error during termination: " + e2, e2);
+            }
+            Exceptions.propagate(e);
```

Comment:
Exceptions should propagate to the blueprint container and prevent it from instantiating the corresponding bean. As a result the bundle should transition in the `Failed` state instead of `Started`. I think bundles depending on services of a failed bundle will enter a 5 minute `GracePeriod` state, and next fail themselves, but it might be they don't get to start at all - not really sure.